File: problems/Java/123. Best Time to Buy and Sell Stock III.java
class Solution {
    public int maxProfit(int[] prices) {
        int b1 = Integer.MIN_VALUE, b2 = Integer.MIN_VALUE;
        int s1 = 0, s2 = 0;
        
        for (int p : prices) {
            b1 = Math.max(b1, -p);
            s1 = Math.max(s1, b1 + p);
            b2 = Math.max(b2, s1 - p);
            s2 = Math.max(s2, b2 + p);
        }
        return s2;
    }
}

File: problems/C++/123. Best Time to Buy and Sell Stock III.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int b1 = INT_MIN, b2 = INT_MIN;
        int s1 = 0, s2 = 0;
        
        for (int p : prices) {
            b1 = max(b1, -p);
            s1 = max(s1, b1 + p);
            b2 = max(b2, s1 - p);
            s2 = max(s2, b2 + p);
        }
        return s2;
    }
};